src/graph/visitor/CheckDepthVisitor.cpp

```diff
@@ -0,0 +1,251 @@
+/* Copyright (c) 2021 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License.
+ */
+
+#include "graph/visitor/CheckDepthVisitor.h"
+
+#include <sstream>
+#include <unordered_map>
+
+#include "common/datatypes/DataSet.h"
+#include "common/datatypes/Edge.h"
+#include "common/datatypes/List.h"
+#include "common/datatypes/Map.h"
+#include "common/datatypes/Path.h"
+#include "common/datatypes/Set.h"
+#include "common/function/FunctionManager.h"
+#include "graph/context/QueryContext.h"
+#include "graph/context/QueryExpressionContext.h"
+#include "graph/context/ValidateContext.h"
+#include "graph/util/SchemaUtil.h"
+#include "graph/visitor/EvaluableExprVisitor.h"
+
+namespace nebula {
+namespace graph {
+
+CheckDepthVisitor::CheckDepthVisitor() {}
+
+void CheckDepthVisitor::visit(ConstantExpression *) {}
+
+void CheckDepthVisitor::visit(UnaryExpression *expr) {
+  checkDepth();
+  SCOPE_EXIT { recoverDepth(); };
+  if (!ok()) return;
+  expr->operand()->accept(this);
+}
+
+void CheckDepthVisitor::visit(TypeCastingExpression *expr) {
+  checkDepth();
+  SCOPE_EXIT { recoverDepth(); };
+  if (!ok()) return;
+  expr->operand()->accept(this);
+  if (!ok()) return;
+}
+
+void CheckDepthVisitor::visit(LabelExpression *) {}
+
+void CheckDepthVisitor::visit(ArithmeticExpression *expr) {
+  checkDepth();
+  SCOPE_EXIT { recoverDepth(); };
+  if (!ok()) return;
+  expr->left()->accept(this);
+  if (!ok()) return;
+  expr->right()->accept(this);
+}
+
+void CheckDepthVisitor::visit(RelationalExpression *expr) {
+  checkDepth();
+  SCOPE_EXIT { recoverDepth(); };
+  if (!ok()) return;
+  expr->left()->accept(this);
+  if (!ok()) return;
+  expr->right()->accept(this);
+}
+
+void CheckDepthVisitor::visit(SubscriptExpression *expr) {
+  checkDepth();
+  SCOPE_EXIT { recoverDepth(); };
+  if (!ok()) return;
```

Comment:
consider refactor to two function that return bool and use: fa() and fb();

Reply:
this pr is closed